### DISCUSSION TITLE
Will this still work on a pi 5?

### DISCUSSION THREAD

**Discussion by PFLAtm** [rust-embedded/rust-raspberrypi-OS-tutorials#216] [Tutorial Questions] [answered]
It was stated in the tutorial that this is for pi3 and pi4 but I guess at this point in time the pi5 wasn't released yet.

**Comment by BartMassey** [comment, marked as answer]
It's going to be more difficult on a Pi 5. The newer processor is quite a bit different, and I'm not sure what the QEMU situation is there. If you do try this on a Pi 5 please do take detailed notes and contribute them back here :grinning: .

**Comment by kindermax** [comment]
Hi, I did some progress porting this kernel to Raspberry Pi 5. 
https://github.com/kindermax/rust-raspberrypi-OS
You can look at commit history on master branch.

However I've stuck on implementing MMU, in particular after the last `isb` barrier it just hangs or the uart stops working - anyway, i can not see any logs.

I tried to replace `isb` with `dsb` barrier, and something works (https://github.com/kindermax/rust-raspberrypi-OS/commit/ffff3300f2b370c85410d89c71038e89d37ccfed) but later in `exceptions` chapter in the `main.rs` where we are trying to access some out of range memory, it does not throws an exception, so my guess that MMU is not working properly.

Also since RPI 5 has all the peripherals on Southbridge RP1 chip, I can not use PL011 uart drivers the usual way without setting up PCIe driver first. Instead I use so called `early uart` which means that some uarts already configured by firmware and are ready to use, no need for driver initialization.
One such uart is debug uart which is 3pin uart, other uart on gpio pins (similar to rpi3/4) can be used with special address (see memory.rs::mmio). In practice I use both uarts so i can load kernel via one uart and debug via another uart

I would like to receive some help/advice so I can move forward)

**Reply by paulgagnon1969** [reply]
Hello kindermax —
│
│ Your July 18 update on this thread was really helpful — especially the MMU isb hang and the early-│ UART / RP1 PCIe wrinkle. We're on a parallel lane at Nexus (startup building a Rust bare-metal │ stack — I nicknamed our lab build "RustOnlyUm"): Pi OS stays up on cores 0–2 for WiFi/SSH, we │ carve core 3 for metal work.
│
│ We got past UART using a similar early-debug path, but we're stuck on the Linux → core 3 handover. │ PSCI CPU_ON returns invalid_params even with an MMU stub + SmpBootData (we borrowed the pattern │ from smallkirby's Pi5 write-up). MMU-off jump from a kmod gets core 3 alive briefly, then the host │ wedges (RCU/watchdog).
│
│ You mentioned swapping isb for dsb got you partial progress (ffff330) but exceptions still look │ wrong. Did you find a clean fix past that ISB barrier on BCM2712 — or a page-table attribute that │ made fault delivery work?
│
│ Happy to swap notes either way — we're logging failures as we go. Thanks for keeping the Pi 5 Rust │ port moving.
│
│ — Paul Gagnon, Nexus Fortified Group
